fix(compiler): use either summary or metadata information when reading from `.d.ts`

tbosch opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

tbosch committed:
fix(compiler): use either summary or metadata information when reading .d.ts files

This fix applies for getting all symbols in file.

mary-poppins commented:
You can preview c5a4e51 at https://pr18912-c5a4e51.ngbuilds.io/.

mary-poppins commented:
You can preview 81b2ef5 at https://pr18912-81b2ef5.ngbuilds.io/.

chuckjaz commented on packages/compiler/src/aot/static_symbol_resolver.ts:
```diff
@@ -251,17 +251,20 @@ export class StaticSymbolResolver {
     this.resolvedSymbols.forEach((resolvedSymbol) => {
       if (resolvedSymbol.symbol.filePath === filePath) {
-        symbols.add(resolvedSymbol.symbol);
+        metadataSymbols.push(resolvedSymbol.symbol);
```
The use of the set was to ensure this array has no duplicates. Is that now ensured some other way?

tbosch replied:
We needed the `Set` as we merged 2 list of symbols. As we are now always using either or, and we know that these lists don't contain duplicates, we are good.

chuckjaz commented on packages/compiler/src/aot/summary_resolver.ts:
```diff
@@ -85,11 +87,12 @@ export class AotSummaryResolver implements SummaryResolver<StaticSymbol> {
+    if (hasSummary != null) {
+      return hasSummary;
     }
-    this.loadedFilePaths.add(filePath);
```
With this moving to after the `if` could potentially allow indirect recursion that was prevented by seeding the cache. Is this is concern?

tbosch replied:
Good point. Will change that.

tbosch replied:
(although this shouldn't happen...)

chuckjaz reviewed (commented):
Some questions.

mary-poppins commented:
You can preview 9d38d7d at https://pr18912-9d38d7d.ngbuilds.io/.

tbosch commented:
Travis is red because of a flake in AIO.

tbosch committed:
fix(compiler): workaround bugs in TS when combining transformers

mary-poppins commented:
You can preview d0218a9 at https://pr18912-d0218a9.ngbuilds.io/.

tbosch committed:
fix(compiler): treat absolute imports as package imports

This is a corner case, and converting them is what
was expected in G3. This also fits the fact that
we already convert package paths into relative paths.

tbosch committed:
build: for bazel, paths in summaries should never contain blaze prefix paths

tbosch committed:
build: correctly collect summary files with bazel

alexeagle commented on packages/bazel/src/ngc-wrapped/index.ts:
```diff
@@ -63,13 +63,30 @@ export function main(args) {
+    throw new Error('rootDirs is not set!');
+  }
+  // When we canonicalize, we want to strip the longest rootDir.
+  rootDirs.sort((a, b) => b.length - a.length);
```
I think it belongs in rules_typescript, let me move it there?

alexeagle replied:
actually this already happens in the compiler_host.ts from rules_typescript, shouldn't need it

alexeagle commented on packages/bazel/src/ngc-wrapped/index.ts:
```diff
@@ -63,13 +63,30 @@ export function main(args) {
     return filePath;
   }
+
+  function relativeToRootDirs(filePath: string): string {
```
let's have one function that takes a parameter 
roots
that defaults to `[tsOptions.rootDir]`
to reduce the duplication?

alexeagle reviewed (approved):
last two commits

mary-poppins commented:
You can preview 0eddfdb at https://pr18912-0eddfdb.ngbuilds.io/.

mary-poppins commented:
You can preview 3dc2c07 at https://pr18912-3dc2c07.ngbuilds.io/.

mary-poppins commented:
You can preview c3740c3 at https://pr18912-c3740c3.ngbuilds.io/.

mary-poppins commented:
You can preview 63ebff6 at https://pr18912-63ebff6.ngbuilds.io/.
